### PR TITLE
CXX-716: use stdx names in the mongocxx via mongocxx/stdx.hpp

### PR DESCRIPTION
**Pull request by Machyne**
switch all `bsconcxx::stdx` to be `mongocxx::stdx` under the mongocxx namespace

### CODE REVIEW

File: src/mongocxx/collection.hpp [hunk 1]
```diff
@@ -329,10 +318,8 @@ class MONGOCXX_API collection {
     ///
     /// @see http://docs.mongodb.org/manual/core/read-operations-introduction/
     ///
-    bsoncxx::stdx::optional<bsoncxx::document::value> find_one(
-        bsoncxx::document::view filter,
-        const options::find& options = options::find()
-    );
+    stdx::optional<bsoncxx::document::value> find_one(
+        bsoncxx::document::view filter, const options::find& options = options::find());
```
Comment (samantharitter) [inline src/mongocxx/collection.hpp]: these closing lines are all clang format changes?

File: src/mongocxx/options/find_one_and_replace.hpp [hunk 2]
```diff
@@ -19,9 +19,11 @@
 #include <cstdint>
 
 #include <bsoncxx/document/view.hpp>
-#include <mongocxx/options/find_one_and_modify.hpp>
 #include <bsoncxx/stdx/optional.hpp>
```
Comment (samantharitter) [inline src/mongocxx/options/find_one_and_replace.hpp]: Do we still need this now?  Should headers that use stdx::optional include <mongocxx/stdx.hpp>, <bsoncxx/stdx/optional.hpp>, or both?

File: src/mongocxx/collection.hpp (hunk 1 above)
Comment (Machyne) [inline src/mongocxx/collection.hpp]: Yes! After removing the `bsoncxx::` clang-format wanted them on the same line. I guess it shrunk just enough? Either that or it was breaking the format before.

File: src/mongocxx/options/find_one_and_replace.hpp (hunk 2 above)
Comment (Machyne) [inline src/mongocxx/options/find_one_and_replace.hpp]: I just checked, and both are necessary to build. The `<bsoncxx/stdx/*>` defines things such as `optional`, and the `<mongocxx/stdx.hpp>` just says to use all of the `bsoncxx::stdx` namespace inside `mongocxx::stdx`.

**Comment by Machyne** [comment]
ping

**Comment by samantharitter** [comment]
LGTM

**Comment by amidvidy** [comment]
LGTM

File: src/mongocxx/client.hpp [hunk 3]
```diff
@@ -164,8 +164,8 @@ class MONGOCXX_API client {
     ///
     /// @return The database
     ///
-    class database database(stdx::string_view name) const &;
-    class database database(stdx::string_view name) const && = delete;
+    class database database(stdx::string_view name) const&;
```
Comment (acmorrow) [inline src/mongocxx/client.hpp]: Seems unrelated, or did clang-format do this?

File: src/mongocxx/stdx.hpp [hunk 4]
```diff
@@ -21,7 +21,7 @@ MONGOCXX_INLINE_NAMESPACE_BEGIN
 namespace stdx {
 
 // We adopt all the bsoncxx polyfills
-using namespace bsoncxx::stdx;
+using namespace ::bsoncxx::stdx;
```
Comment (acmorrow) [inline src/mongocxx/stdx.hpp]: Nice improvement

File: src/mongocxx/write_concern.hpp [hunk 5]
```diff
@@ -23,6 +23,7 @@
 
 #include <bsoncxx/stdx/optional.hpp>
 #include <bsoncxx/stdx/string_view.hpp>
+
```
Comment (acmorrow) [inline src/mongocxx/write_concern.hpp]: Is this change following style elsewhere? I don't remember that we separated the blocks for bsoncxx and mongocxx headers as a general principle, I think we just sorted them together,

File: src/mongocxx/options/find_one_and_replace.hpp (hunk 2 above)
Comment (acmorrow) [inline src/mongocxx/options/find_one_and_replace.hpp]: That may be more of an oversight than a design choice. Can anyone think of a good reason not to include the bsoncxx/stdx headers inside the mongocxx/stdx,hpp header?

File: src/mongocxx/stdx.hpp (hunk 4 above)
Comment (acmorrow) [inline src/mongocxx/stdx.hpp]: When I wrote this file, I felt a little bad. I wondered if it would have been better to have mongocxx/stdx/optional.hpp, etc., headers, but it just seemed like massive overkill.

However, I do wonder if we should be more restrictive here, and just import name by name as needed by mongocxx.

In other words, rather than saying using namespace in a header (generally not a great idea), just say

using ::bsoncxx::stdx::optional;
using ::bsoncxx::stdx::string_view;

etc.

What do people think?

File: src/mongocxx/options/find_one_and_replace.hpp (hunk 2 above)
Comment (amidvidy) [inline src/mongocxx/options/find_one_and_replace.hpp]: i think this was an explicit choice to avoid facade headers

**Comment by acmorrow** [comment]
I see that this got merged already, which is fine. I do have a few small comments and suggestions above. 
The only real change I see as potentially needed is to include the bsoncxx/stdx headers into mongocxx/stdx.hpp and import type by type with usings.

File: src/mongocxx/options/find_one_and_replace.hpp (hunk 2 above)
Comment (acmorrow) [inline src/mongocxx/options/find_one_and_replace.hpp]: Yeah. Generally a principle I support but maybe working against us here. If we are really only including the bsoncxx/stdx/xyz.hpp headers we need and then individually pulling in the types with using, I think it is something we could live with.
